sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java

```diff
@@ -120,6 +122,40 @@ private static String nextAssertionName() {
   // Do not instantiate.
   private PAssert() {}
 
+  private static final class DefaultConcludeFn extends DoFn<SuccessOrFailure, Void> {
+
+    private Counter successCounter;
+    private Counter failureCounter;
+
+    @StartBundle
+    public void startBundle(Context c) {
+      successCounter = Metrics.counter(PAssert.class, PAssert.SUCCESS_COUNTER);
+      failureCounter = Metrics.counter(PAssert.class, PAssert.FAILURE_COUNTER);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      SuccessOrFailure e = c.element();
+      if (e.isSuccess()) {
+        successCounter.inc();
+      } else {
+        failureCounter.inc();
+        throw e.assertionError();
+      }
+    }
+  }
+
+  private static final PTransform<? super PCollection<SuccessOrFailure>, ? super PCollection<Void>>
```

Comment:
I think to make it easy to use the graph manipulations, this should be a named composite transform. Something like:

```
static DefaultConcludeTransform extends PTransform<PCollection<SuccessOrFailure>, PCollection<Void>> {
  public PCollection<Void> expand(PCollection<SuccessOrFailure> input) {
    return input.apply(ParDo.of(new DefaultConcludeFn());
  }
}
```